### PR TITLE
ALSA/ASoC: use pm_runtime_resume_and_get() instead of pm_runtime_get_sync()

### PR DESCRIPTION
**Pull request by plbossart**
this PR extends the changes done for SOF and SoundWire to the rest of the sound/ tree.

the patches are ordered with different level of changes:
a) no flow changes
b) addition of -EACCES filter
c) addition of error handling
d) rename adding release of reference on error - but no tests on errors

### CODE REVIEW

File: sound/hda/hdac_device.c [hunk 1]
```diff
@@ -566,7 +566,13 @@ EXPORT_SYMBOL_GPL(snd_hdac_get_connections);
  */
 int snd_hdac_power_up(struct hdac_device *codec)
 {
-	return pm_runtime_get_sync(&codec->dev);
+	int ret;
+
+	ret = pm_runtime_resume_and_get(&codec->dev);
+	if (ret < 0 && ret != -EACCES)
```
Comment (ujfalusi) [inline sound/hda/hdac_device.c]: This comment applies to other, similar patches:
- commit message have `-ENOACCES`, but it is `-EACCES`
- Why -EACCES is success and not error? It was treated as error before...
- `pm_runtime_resume_and_get()` return 0 on success or the error code on failure
```c
if (ret && ret != -EACCES)
	return ret;
```
or
```c
if (ret == -EACCES)
	ret = 0;

return ret;
```
might be better if -EACCES _is_ really not an error.

Comment (plbossart) [reply sound/hda/hdac_device.c]: -EACCES is returned by pm_runtime is disabled or not allowed. It's not an error. Ideally it should have been filtered at the source in the pm_runtime code.

if (ret && ret != -EACCES)
	return ret;

isn't very different from my suggestion

if (ret < 0 && ret != -EACCES)
	return ret;

I prefer to test for negative values explicitly.

File: sound/soc/soc-component.c [hunk 2]
```diff
@@ -1213,11 +1213,11 @@ int snd_soc_pcm_component_pm_runtime_get(struct snd_soc_pcm_runtime *rtd,
 	int i;
 
 	for_each_rtd_components(rtd, i, component) {
-		int ret = pm_runtime_get_sync(component->dev);
-		if (ret < 0 && ret != -EACCES) {
-			pm_runtime_put_noidle(component->dev);
+		int ret = pm_runtime_resume_and_get(component->dev);
```
Comment (ranj063) [inline sound/soc/soc-component.c:1216]: @plbossart I think this is not just simplifying but also changing the flow. pm_runtime_resume_and_get() would also fo a put_noidle() in case of -EACCES

Comment (ranj063) [reply sound/soc/soc-component.c:1216]: OK this change is correct but we should clarify that its not just a simplification but also a correction for the incorrect error path flow.

Comment (plbossart) [reply sound/soc/soc-component.c:1216]: done

File: sound/hda/hdac_device.c (hunk 1 above)
Comment (plbossart) [reply sound/hda/hdac_device.c]: fixed the commit messages, see update

**Review by ranj063 — DISMISSED** [review]

File: sound/hda/hdac_device.c (hunk 1 above)
Comment (ujfalusi) [reply sound/hda/hdac_device.c]: Ok, but 'pm_runtime_resume_and_get()' will never return positive number, testing for 'ret < 0' gives the impression that a positive return number is possible and it is a success indicator.

Comment (ranj063) [reply sound/hda/hdac_device.c]: @ujfalusi this patch has been removed but even otherwise testing for ret < 0 makes sense. just ret != -EACCESS would include 0 as well

Comment (ujfalusi) [reply sound/hda/hdac_device.c]: @ranj063, oh, ok.
I was suggesting:
```c
if (ret && ret != -EACCES)
	return ret;

return 0;
```
or
```c
if (ret == -EACCES)
	ret = 0;

return ret;
```

Comment (plbossart) [reply sound/hda/hdac_device.c]: @ujfalusi there's already a precedent for using < 0 across drivers when using pm_runtime_get_sync, let's keep the same style as before....

Comment (ujfalusi) [reply sound/hda/hdac_device.c]: Right, it looks that way. Almost all `pm_runtime_resume_and_get()` is with `if (ret < 0)`.
In those cases the error check was not updated when converting to the simplified wrapper.

In case of `pm_runtime_get_sync()` it is valid as it can return positive number which is not an error, but the resume_and_get wrapper got it wrong (arguably).

Comment (ujfalusi) [reply sound/hda/hdac_device.c]: I stand corrected, `pm_runtime_resume_and_get()` does things correctly, it just wraps __pm_runtime_resume() and makes it return 0 or error code and removes the positive number as return value.

Eating up -EACCES does change the behavior as we did not done that in the past.

Comment (plbossart) [reply sound/hda/hdac_device.c]: For my education, where did you see that pm_runtime_get_sync() can return a positive value @ujfalusi 

Also if you're good with the code as is please approve or reject if you're not, I am not sure where you stand on this PR.

Comment (ujfalusi) [reply sound/hda/hdac_device.c]: @plbossart, `pm_runtime_get_sync()` is a wrapper for `__pm_runtime_resume()`, which is calling `rpm_resume()` which returns 0 on success, positive number if the status was already enabled.
pm_runtime_get_sync:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/power/runtime_pm.rst#n385
and pm_runtime_resume:
https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/Documentation/power/runtime_pm.rst#n338

The `pm_runtime_resume_and_get()` is returning 0 if `retval >= 0` so the code does not need to check for positive numbers, `!ret` is success `ret != 0` is error.

Let me check the PR again.

File: sound/soc/ti/davinci-mcasp.c [hunk 3]
```diff
@@ -2111,8 +2112,11 @@ static int davinci_mcasp_gpio_request(struct gpio_chip *chip, unsigned offset)
 	}
 
 	/* Do not change the PIN yet */
+	ret = pm_runtime_resume_and_get(mcasp->dev);
+	if (ret < 0 && ret != -EACCES)
+		return ret;
```
Comment (ujfalusi) [inline sound/soc/ti/davinci-mcasp.c]: Please do not hide the `-EACCES`, it means that the the device remained _disabled_ because the `rpm_resume()` did not run because the disable_depth was different from 0.
We can only return 0 if the McASP _is_ enabled.

File: sound/soc/codecs/tas2552.c [hunk 4]
```diff
@@ -581,8 +581,8 @@ static int tas2552_component_probe(struct snd_soc_component *component)
 
 	gpiod_set_value(tas2552->enable_gpio, 1);
 
-	ret = pm_runtime_get_sync(component->dev);
-	if (ret < 0) {
+	ret = pm_runtime_resume_and_get(component->dev);
+	if (ret < 0 && ret != -EACCES) {
```
Comment (ujfalusi) [inline sound/soc/codecs/tas2552.c]: Same comment as for McASP: if we get `-EACCES` it means that the device is _not_ enabled...

Comment (plbossart) [reply sound/soc/codecs/tas2552.c]: @ujfalusi the -EACCESS means pm_runtime is disabled, not that the device is disabled. so the device is active and never went to low-power state.

````
 int pm_runtime_resume(struct device *dev);
    - execute the subsystem-level resume callback for the device; returns 0 on
      success, 1 if the device's runtime PM status was already 'active' or
      error code on failure, where -EAGAIN means it may be safe to attempt to
      resume the device again in future, but 'power.runtime_error' should be
      checked additionally, and -EACCES means that 'power.disable_depth' is
      different from 0
`````

In other words, it's not an error and we can keep going.

Comment (ujfalusi) [reply sound/soc/codecs/tas2552.c]: @plbossart, right, I did interpreted it wrong, you are correct. If the pm_runtime_enable() is not called on the device then yes, -EACCES could be returned, but it is called in probe time, so for drivers itself this will never happen, unless the pm_runtime_enable() is conditional in some way, but I don't see that in the tas2552.c nor in davinci-mcasp.c for example.

In framework level this make sense as there might be drivers managed by the framework which does not use runtime_pm.

Comment (plbossart) [reply sound/soc/codecs/tas2552.c]: @ujfalusi I added -EACCES by default in all cases, because it's hard to figure out if it's missing by mistake or intentionally (because it cannot happen by design).

However I did confuse pm_runtime_allow/forbid (which is accessible to the user via the 'control' sysfs file) and pm_runtime_enable/disable, which can only be done by drivers. You're absolutely right that -EACCES can only be returned if pm_runtime_enable is missing somehow.

So yes I agree that we should more carefully re-inspect all the -EACCES additional checks when there were missing. I will assume that if it was present before then we leave it as is. 

I will revisit all this more carefully, good feedback

**Comment by plbossart** [comment]
ADLP_RVP_NOCODEC timeout in https://sof-ci.01.org/linuxpr/PR3663/build182/devicetest/

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by ujfalusi** [comment]
@plbossart, still too many TIMEOUTs?

**Comment by plbossart** [comment]
> @plbossart, still too many TIMEOUTs?

3 TIMEOUTs, and on different devices. It's becoming really hard to debug...

**Comment by plbossart** [comment]
SOFCI TEST

**Comment by plbossart** [comment]
It's really not so bad @ujfalusi, see https://sof-ci.01.org/linuxpr/PR3663/build194/devicetest/, this time there's only a FAIL that's due to a bad 'sync', looks harmless, and the usual issues with suspend-resume on WHL UP board.

**Review by ujfalusi — APPROVED** [review]

**Comment by plbossart** [comment]
@ranj063 @bardliao @dbaluta  @kv2019i  can you help review?

File: sound/soc/codecs/arizona-jack.c [hunk 5]
```diff
@@ -238,7 +238,7 @@ static void arizona_start_mic(struct arizona_priv *info)
 	unsigned int mode;
 
 	/* Microphone detection can't use idle mode */
-	pm_runtime_get_sync(arizona->dev);
+	pm_runtime_resume_and_get(arizona->dev);
```
Comment (kv2019i) [inline sound/soc/codecs/arizona-jack.c]: Hmm. @plbossart is this safe usage when replacing get_sync where return value is not checked. After this change, usage count is not increased on error, but execution is still continued and there's an eventual unconditional call to pm_runtime_put(). I don't see __pm_runtime_suspend() in power.c handling this case gracefully.

**Review by kv2019i — CHANGES_REQUESTED** [review]
90% looks good, but the "no return value check" cases are a bit suspicious.

File: sound/soc/codecs/arizona-jack.c (hunk 5 above)
Comment (plbossart) [reply sound/soc/codecs/arizona-jack.c]: You're not wrong @kv2019i, maybe we shouldn't touch this sort of code. this doesn't seem to good to me...

````c
	ret = regmap_update_bits_check(arizona->regmap, ARIZONA_MIC_DETECT_1,
				       ARIZONA_MICD_ENA, ARIZONA_MICD_ENA,
				       &change);
	if (ret < 0) {
		dev_err(arizona->dev, "Failed to enable micd: %d\n", ret);
	} else if (!change) {
		regulator_disable(info->micvdd);
		pm_runtime_put_autosuspend(arizona->dev);
	}
````

File: sound/soc/ti/omap-dmic.c [hunk 6]
```diff
@@ -417,7 +417,7 @@ static int omap_dmic_probe(struct snd_soc_dai *dai)
 	pm_runtime_enable(dmic->dev);
 
 	/* Disable lines while request is ongoing */
-	pm_runtime_get_sync(dmic->dev);
+	pm_runtime_resume_and_get(dmic->dev);
```
Comment (ujfalusi) [inline sound/soc/ti/omap-dmic.c]: @plbossart, in light of @kv2019i's comment, this is also on the not that correct side?

Comment (plbossart) [reply sound/soc/ti/omap-dmic.c]: Yes, I've removed all the cases where there's not a test already. It's just nuts to try to fix all this.

**Comment by plbossart** [comment]
@ujfalusi @kv2019i @bardliao @ranj063 I removed all the controversial cases, I didn't think this would be that complicated/broken...Also made sure all the allmodconfig W=1 compilation works, I had typos in my previous version

**Review by bardliao — DISMISSED** [review]

File: sound/x86/intel_hdmi_audio.c [hunk 7]
```diff
@@ -1802,8 +1804,11 @@ static int __hdmi_lpe_audio_probe(struct platform_device *pdev)
 	pdata->notify_audio_lpe = notify_audio_lpe;
 	spin_unlock_irq(&pdata->lpe_audio_slock);
 
+	pm_runtime_set_autosuspend_delay(&pdev->dev, INTEL_HDMI_AUDIO_SUSPEND_DELAY_MS);
```
Comment (kv2019i) [inline sound/x86/intel_hdmi_audio.c:1813]: I have really no idea about this patch. Maybe should be a separate patch and not in this series?

**Review by kv2019i — APPROVED** [review]
The lpe HDMI driver change pops up a bit in the series. Probably ok, but it's a different type of change and harder to review (is 5000ms a reasonable value or not, will something break if runtime-pm suddently starts to be used when it before was not). But for everything else, good to go.

File: sound/x86/intel_hdmi_audio.c (hunk 7 above)
Comment (plbossart) [reply sound/x86/intel_hdmi_audio.c:1813]: yes, this would not be an ASoC patch and will be send separately. I removed the earlier HDaudio patches which don't work at all with pm_runtime_resume_and_get(), so now these Atom HDMI patches seem a bit out of place indeed.

**Review by ranj063 — DISMISSED** [review]

**Review by ujfalusi — APPROVED** [review]

**Comment by plbossart** [comment]
Sorry @bardliao @kv2019i @ujfalusi @ranj063 there was an upstream conflict, and after the rebase your reviews were lost. Sorry for the inconvenience.

**Review by ujfalusi — APPROVED** [review]

**Review by bardliao — APPROVED** [review]

**Review by kv2019i — APPROVED** [review]

**Review by ranj063 — APPROVED** [review]
